From f396e23cda825ad4785a7e5e97e6dd462dfc5035 Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH 41/67] nuplayer: Allow devices to force a flush when surface
 changes

This fixes some playback problems on the Zenfone2 (running on 5.0
based blobs).

The problems are easily reproduced by viewing a video in Chrome and
then switching to full-screen mode.  The change in surface causes
the log to get flooded with messages like:

 dc_device_post: Buffer ID=14460 not imported by D

until Android crashes.  With this change, it ends up correctly
mapping the surfaces after the flush fixing the problem.

The code that is being #ifdef'ed out was added in M in commit
 nuplayer: use codec->setSurface when possible to avoid seeking

and it appears that our blobs simply cannot handle this optimization
and therefore we need a means by which to disable it.

Change-Id: I1dd915caec14be6ffb28f708f64b41ebb5a44cd2
---
 media/libmediaplayerservice/nuplayer/Android.bp   | 1 +
 media/libmediaplayerservice/nuplayer/NuPlayer.cpp | 5 ++++-
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/media/libmediaplayerservice/nuplayer/Android.bp b/media/libmediaplayerservice/nuplayer/Android.bp
index 645bb7a90..62ffc574b 100644
--- a/media/libmediaplayerservice/nuplayer/Android.bp
+++ b/media/libmediaplayerservice/nuplayer/Android.bp
@@ -33,6 +33,7 @@ cc_library_static {
     cflags: [
         "-Werror",
         "-Wall",
+        "-DCANNOT_SET_SURFACE_WITHOUT_A_FLUSH",
     ],
 
     product_variables: {
diff --git a/media/libmediaplayerservice/nuplayer/NuPlayer.cpp b/media/libmediaplayerservice/nuplayer/NuPlayer.cpp
index a5f5fc675..795bc3af2 100644
--- a/media/libmediaplayerservice/nuplayer/NuPlayer.cpp
+++ b/media/libmediaplayerservice/nuplayer/NuPlayer.cpp
@@ -771,7 +771,10 @@ void NuPlayer::onMessageReceived(const sp<AMessage> &msg) {
             // When mStarted is true, mSource must have been set.
             if (mSource == NULL || !mStarted || mSource->getFormat(false /* audio */) == NULL
                     // NOTE: mVideoDecoder's mSurface is always non-null
-                    || (mVideoDecoder != NULL && mVideoDecoder->setVideoSurface(surface) == OK)) {
+#ifndef CANNOT_SET_SURFACE_WITHOUT_A_FLUSH
+                    || (mVideoDecoder != NULL && mVideoDecoder->setVideoSurface(surface) == OK)
+#endif
+                ) {
                 performSetSurface(surface);
                 break;
             }
-- 
2.11.0

